src/main/scala/corpus.scala

```diff
@@ -0,0 +1,87 @@
+package treesitter.lsp
+
+import parsley.character.*
+import parsley.debug.*
+import parsley.combinator.*
+import parsley.Parsley
+
+enum LispNode:
+  case Leaf(name: String)
+  case Nest(title: String, nodes: List[LispNode])
+
+object LispNode:
+  export parsers.parser
+
+  private object parsers:
+    def constructNode(nm: String, subnodes: List[LispNode]) =
+      subnodes match
+        case head :: next => LispNode.Nest(nm, subnodes)
+        case Nil          => LispNode.Leaf(nm)
+
+    val nodeName = some(letter <|> char('_')).map(_.mkString)
+
+    extension [A](p: => Parsley[A]) def wsp = p <* whitespaces
+
+    lazy val parser: parsley.Parsley[LispNode] = (
+      char('(') *>
+        (nodeName.wsp <~> {
+
+          val rec = many(parser)
+
+          val simple = nodeName
+            .map(LispNode.Leaf(_))
+            .map(List(_))
+
+          simple | rec
+
+        }).map(constructNode) <*
```

Comment:
`(p <~> q).map(f)` is not idiomatic, either:

* `lift2(f, p, q)`
* `f.lift(p, q)` (with `parsley.implicits.lift.Lift2`)
* `(p, q).zipped(f)` (with `parsley.implicits.zipped.Zipped2`), this has better inference than the previous two!
* or use my above suggestion and use the generic bridge traits to instead write `LispNode(nodeName, { val rec = ... })`

Reply:
The forward thinking consideration would be if you want to add position information into the nodes. This becomes less invasive if you're using the bridge option (the bridge abstracts the position tracking) compared with the `zipped` one. I agree that `zipped` is usually the way to go compared with `lift` unless the functions behave themselves.